test/src/console/foreign.spec.ts

```diff
@@ -0,0 +1,302 @@
+} from '../../../lib/console/foreign';
+
+import {
+  CodeCellModel, CodeCellWidget
+} from '../../../lib/notebook/cells';
+
+import {
+  defaultRenderMime
+} from '../utils';
+
+
+const relevantTypes = [
+  'execute_input',
+  'execute_result',
+  'display_data',
+  'stream',
+  'error',
+  'clear_output'
+].reduce((acc, val) => {
+  acc.add(val);
+  return acc;
+}, new Set<string>());
+
+
+class TestHandler extends ForeignHandler {
+
+  readonly injected: ISignal<this, void>;
+
+  readonly received: ISignal<this, void>;
+
+  readonly rejected: ISignal<this, void>;
+
+  methods: string[] = [];
+
+  dispose(): void {
+    if (this.isDisposed) {
+      return;
+    }
+    super.dispose();
+    clearSignalData(this);
+  }
+
+  protected onIOPubMessage(sender: Kernel.IKernel, msg: KernelMessage.IIOPubMessage): boolean {
+    let injected = super.onIOPubMessage(sender, msg);
+    this.received.emit(void 0);
+    if (injected) {
+      this.injected.emit(void 0);
+    } else {
+      // If the message was not injected but otherwise would have been, emit
+      // a rejected signal. This should only happen if `enabled` is `false`.
+      let session = (msg.parent_header as KernelMessage.IHeader).session;
+      let msgType = msg.header.msg_type;
+      if (session !== this.kernel.clientId && relevantTypes.has(msgType)) {
+        this.rejected.emit(void 0);
+      }
+    }
+    return injected;
+  }
+}
+
+
+defineSignal(TestHandler.prototype, 'injected');
+defineSignal(TestHandler.prototype, 'received');
+defineSignal(TestHandler.prototype, 'rejected');
+
+
+const rendermime = defaultRenderMime();
+const renderer: ForeignHandler.IRenderer = {
+  createCell: () => {
+    let renderer = CodeMirrorConsoleRenderer.defaultCodeCellRenderer;
+    let cell = new CodeCellWidget({ rendermime, renderer });
+    cell.model = new CodeCellModel();
+    return cell;
+  }
+};
+
+
+describe('console/foreign', () => {
+
+  describe('ForeignHandler', () => {
+
+    describe('#constructor()', () => {
+
+      it('should create a new foreign handler', () => {
+        let handler = new ForeignHandler({
+          kernel: null,
+          parent: null,
+          renderer
+        });
+        expect(handler).to.be.a(ForeignHandler);
+      });
+
+    });
+
+    describe('#enabled', () => {
+
+      it('should default to `true`', () => {
+        let handler = new ForeignHandler({
+          kernel: null,
+          parent: null,
+          renderer
+        });
+        expect(handler.enabled).to.be(true);
+      });
+
+      it('should allow foreign cells to be injected if `true`', done => {
+        let path = utils.uuid();
+        let code = 'print("#enabled:true")';
+        let parent = new Panel();
+        Promise.all([
+          Session.startNew({ path }), Session.startNew({ path })
+        ]).then(([a, b]) => {
+          let handler = new TestHandler({ kernel: a.kernel, parent, renderer });
+          handler.injected.connect(() => {
+            a.dispose();
```

Comment:
disposing of a session does not shut it down on the server, these will be left open.